lisp/ess-developer.el

```diff
@@ -377,21 +430,81 @@ If ALL is non-nil, deactivate in all open R buffers."
                      (equal pack package)))
         (ess-developer -1)))))
 
-(defun ess-developer-load-package ()
-  "Interface to load_all function from devtools package."
-  (interactive)
+(defun ess-developer--get-process-package-path ()
+  (with-ess-process-buffer t
+    (when (boundp 'ess-developer-local-package)
+      (cdr ess-developer-local-package))))
+
+(defun ess-developer-send-process (command &optional msg)
   (ess-force-buffer-current)
-  (let ((package (ess-developer--get-package-path)))
-    (unless (and package ess-developer)
-      ;; ask only when not obvious
-      (setq package
-            (read-directory-name "Package: " package nil t nil)))
-    (unless (file-exists-p (expand-file-name ess-developer-root-file package))
+  (let* ((path (or (cdr ess-developer-local-package)
+                   (ess-developer--get-package-path)))
+         (name (when path
+                 (file-name-nondirectory (directory-file-name path)))))
+    ;; Ask package directory only when not obvious
+    (unless (or ess-developer-local-package path)
+      (setq path (read-directory-name "Package: " path nil t nil)))
+    (unless (file-exists-p (expand-file-name ess-developer-root-file path))
       (error "Not a valid package. No '%s' found in `%s'."
-             ess-developer-root-file package))
-    (message "Loading %s" (abbreviate-file-name package))
-    (ess-eval-linewise
-     (format ess-developer-load-package-command package))))
+             ess-developer-root-file path))
+    (message msg name)
+    (let ((name (or (car ess-developer-local-package)
+                    (ess-developer--get-package-name path))))
+      (ess-developer--init-process-local-vars name path))
+    (ess-eval-linewise (format command path))))
+
+(defun ess-developer--init-process-local-vars (name path)
+  (with-ess-process-buffer nil
+    (setq-local ess-developer-local-package (cons name path))))
+
+(defun ess-developer-load-package ()
+  "Interface to load_all() function from devtools package.
+
+Without prefix, load the package. With single prefix, recompile
+before loading. With double prefix, unload the package."
+  (interactive)
+  (cond ((equal current-prefix-arg '(16))
+         (ess-developer-unload-package))
+        ((equal current-prefix-arg '(4))
+         (ess-developer-send-process "devtools::load_all('%s', recompile = TRUE)\n"
+                                     "Recompiling %s"))
+        (t
+         (ess-developer-send-process "devtools::load_all('%s')\n"
+                                     "Loading %s"))))
+
+(defun ess-developer-unload-package ()
+  "Interface to unload() function from devtools package."
+  (interactive)
+  (ess-developer-send-process "devtools::unload('%s')\n"
+                              "Unloading %s"))
+
+(defun ess-developer-check-package ()
```

Comment:
My common operation is to check without the vignette. I poropose that we have an allist of possible check commands with the default being the first one in the list. It's one extra RET after C-u but it saves us a lot of trouble. 

A major benefit is that users can add their own check commands with whatever additional build args they need.

It might be good to extract all devtools commands into user vars for UI consistency.